src/modules/module_25400.c

```diff
@@ -483,15 +478,81 @@ int module_hash_decode (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSE
   return (PARSER_OK);
 }
 
-int module_hash_encode (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const void *digest_buf, MAYBE_UNUSED const salt_t *salt, MAYBE_UNUSED const void *esalt_buf, MAYBE_UNUSED const void *hook_salt_buf, MAYBE_UNUSED const hashinfo_t *hash_info, char *line_buf, MAYBE_UNUSED const int line_size)
+int module_build_plain_postprocess (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const hashes_t *hashes, MAYBE_UNUSED const void *tmps, const u32 *src_buf, MAYBE_UNUSED const size_t src_sz, MAYBE_UNUSED const int src_len, u32 *dst_buf, MAYBE_UNUSED const size_t dst_sz)
 {
-  const pdf_t *pdf = (const pdf_t *) esalt_buf;
+  const u32 padding[8] =
+  {
+    0x5e4ebf28,
+    0x418a754e,
+    0x564e0064,
+    0x0801faff,
+    0xb6002e2e,
+    0x803e68d0,
+    0xfea90c2f,
+    0x7a695364
+  };
+
+  pdf14_tmp_t *pdf_tmp = (pdf14_tmp_t *) tmps;
+  pdf_t *pdf = (pdf_t *) hashes->esalts_buf;
 
+  // if the password in tmp->out is equal to the padding, then we recovered just the owner-password
+  if(pdf_tmp->out[0]==padding[0] && pdf_tmp->out[1]==padding[1] && pdf_tmp->out[2]==padding[2] && pdf_tmp->out[3]==padding[3])
+  {
+    return snprintf ((char *) dst_buf, dst_sz, "%s    (user password not set)", (char *) src_buf);
+  }
+
+  // cast out buffer to byte such that we can do a byte per byte comparison
+  u32 *u32OutBufPtr = pdf_tmp->out;
+  u8 *u8OutBufPtr;
+  u8OutBufPtr = (u8*) u32OutBufPtr;
+
+  // cast padding buffer to byte such that we can do a byte per byte comparison
+  const u32 *u32OutPadPtr = padding;
+  const u8 *u8OutPadPtr;
+  u8OutPadPtr = (u8*) u32OutPadPtr;
+
+  bool remove_padding=false;
+  int i_padding=0;
+  for(int i=0;i<16;i++)
+  {
+    if(u8OutBufPtr[i]==u8OutPadPtr[i_padding] || remove_padding)
+    {
+    u8OutBufPtr[i]=0x0;
+    remove_padding=true;
+    }
+  }
+
+  // if the password in tmp->out is equal to the password tried, then we recovered just the owner-password or just the user-password
+  //   we check whether we already have a user-password in the hash
+  //   TODO would be better to actually also verify the u-value whether we've retrieved the correct user-password,
+  //     however, we'd need to include a lot of code/complexity here to do so (or call into 10500 kernel).
+  //     this seems relevant: run_kernel (hashcat_ctx, device_param, KERN_RUN_3, 0, 1, false, 0)
```

Comment:
is it possible to call into a kernel from a kernel? 

any other kernel already doing that?

Reply:
Please see comment below